docs/Hooks.md

```diff
@@ -1,16 +1,60 @@
+The difference between these two functions is in the arguments they take to construct the respective hook. To define a pure hook we only need to specify the phase and the hook function to run in that phase, you can read more about the phases and the signatures in the next section, but an example of a simple hook might be
+```coffee
+sbvrUtils.addHook 'method'. 'vocabulary', 'resource',
+	POSTRUN: ({ req, request, tx }) ->
+		#Hook logic
+```
 
-In order to to roll back the transaction you can either throw an error, or return a rejected promise. Also, any promises that are returned will be waited on before continuing with processing the request.
+Hooks will have access to a `tx` object representing the transaction of the current request, in order to to roll back the transaction you can either throw an error, or return a rejected promise.  Also, any promises that are returned will be waited on before continuing with processing the request.
+However, some hooks might need to perform actions against external resources, such as HTTP calls to different services or other forms of side-effectful action. To undo these action we can not rely on the `tx` object and must take to setup the appropriate rollback logic, should the request error out; remember this can also happen at a later time than the hook runs, e.g. we can perform some `PRERUN` action, only to realise the request fails when we later attempt to run it, because of some database constraint.
 
-## Hook points
+To deal with these cases we can define a side-effect hook, the signature is only slightly different:
+```coffee
+sbvrUtils.addHook 'method'. 'vocabulary', 'resource',
+	POSTRUN:
+		RUN: ({ req, request, registerRollback }) ->
+			#Hook logic
+		ROLLBACK: (arg1, arg2) ->
+			#Rollback logic
+```
+
+The hook will now be passed the `registerRollback` function which can be used to store any information we need to perform the rollback (e.g. the ID of a resource we must delete). We can use registerRollback to supply these arguments, which will be later passed to the supplied `ROLLBACK` function if we need to undo the side-effects of the hook.
+`registerRollback` may be called once with all the arguments we intend to supply, or multiple times by passing each argument as soon as it becomes available. The arguments will be passed to the `ROLLBACK` function in the order they have been received.
+
+We can take advantage of this by writing the `ROLLBACK` function so that we can perform partial rollback of our side-effects, as such:
+```coffee
+PHASE:
+	RUN: ({ ..., registerRollback }) ->
+		createResource(1)
```

Comment:
It might be worth noting that this is more a case of `createExternalResource` because if you use the hook's api/tx objects to do the creation then it will automatically be rolled back